### PR TITLE
Closes #1000: Export IIS metrics to prometheus pushgateway

### LINKED ISSUE #1000: Export IIS metrics to prometheus pushgateway
Currently IIS execution report including various metrics is generated during IIS exporting phase and is made available as JSON file and via the oozie API after IIS workflow is completed.

One problem with this approach is the fact we need to wait until IIS finishes its job to know the metrics and potentially to spot anomalies. Some problems could be spotted at pretty early stage, e.g. during IIS import, so it would be reasonable to export metrics as soon as they were generated to avoid losing precious time.

We could take advantage of already existing prometheus and grafana instances to gather and present IIS metrics. Prometheus push gateway was already deployed and is ready to be used by IIS workflows. Having IIS hooked up with prometheus would expand whole new set of capabilities such as spotting regression based on an insight into the history of IIS executions, alerting, graphical presentation etc.

Since each module already covered with reporting functionality produces HDFS datastore holding [`ReportEntry`](https://github.com/openaire/iis/blob/master/iis-schemas/src/main/avro/eu/dnetlib/iis/common/ReportEntry.avdl) avro records we could quite easily provide generic action relying on those records (coming from one or more modules) and sending them to prometheus push gateway. In the first iteration we could place such action after each IIS main phase: import, processing, export. This way we could e.g. spot import related anomalies just after import phase finishes. Later on we could consider introducing more exporting checkpoints what will result in decreasing the delay between metrics generation time and their availability in prometheus.

We should agree on a naming convention of exported metrics, probably by relying on current report entry name, taking an environment into account (beta vs production, we could rely on user name triggering IIS job: either `dnet.beta` or `dnet.production`). Also report entry type (`counter` or `duration`) could be distinguished.

New solution should be implemented and defined in `iis-wf-report` module and it should be extremely fault tolerant (any error during pushing the results to pushgateway should not result in job interruption). Already implemented reporting mechanism could be still operating alongside.

### PR DESCRIPTION
**Pull request by przemyslawjacewicz**
### Overview
This PR adds functionality of pushing IIS reports to prometheus pushgateway.

Two types of metrics are created
 * for counter report entries metrics are created by replacing dots with underscores with additional support for labels
 * for duration report entries metrics are created by replacing dots with underscores and adding '_seconds' suffix

 For testing needs pushgateway pushing is replaced by persisting metrics to file with additional conversion to report entries. 

### Deployment instructions
This PR results in one new required workflow property
 * `metric_pusher_address` - the address of pushgateway service
 
### Notes
This PR additionally includes an update of report generation testing for primary processing integration test - addition of report generation test for
* affiliation matching dedup
* affiliation matching project based
* doc community
* doc aoftware url webcrawl

### CODE REVIEW

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        };
+        return create(pushGatewaySupplier);
+    }
+
+    public MetricPusher create(Supplier<PushGateway> pushGatewaySupplier) {
+        return (collectorRegistry, jobName) -> Optional
+                .ofNullable(pushGatewaySupplier.get())
+                .ifPresent(pushGateway -> {
+                    try {
+                        pushGateway.pushAdd(collectorRegistry, jobName);
```
Comment (mpol) [inline iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java:32]: Should we be using `pushAdd`? It replaces existing metrics with the same name, but if some are missing in a report we will be left with old values. I would suggest rather using `push` here.

I also think job name should always be "iis" and "primary_import" etc. would better fit an additional grouping key named eg. `wf`.

We also need a label to distinguish runs for production, beta, etc. I think it can be a grouping key too, named `env`, like the label we use for services.

Comment (mpol) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java:32]: From today's call:
-  let's use `user` not `env` as a grouping key
- don't use `wf`, group just by `user` - this way we can export all the available reports every time

**Comment by mpol** [comment]
A hard case for better labelled metrics. It would be very nice to have:
- `import.content.url.bytype.pdf` => `import_content_url{type="pdf"}`
- `import.content.pdf.correct` => `import_content{type="pdf",status="correct"}`
- `import.content.pdf.invalid.sizeExceeded` => `import.content{type="pdf",status="sizeExceeded"}`

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                .collect(Collectors.groupingBy(x -> x.metricName)).entrySet().stream()
+                .flatMap(entry -> {
+                    String metricName = entry.getKey();
+                    List<MetricNameAndLabelValuesAndValue> metricNameAndLabelValuesAndValueList = entry.getValue();
+                    return Optional
+                            .ofNullable(labelNamesByMetricName.get(metricName))
+                            .map(labelNames -> Stream
+                                    .of(buildGaugeWithLabels(metricName, help, labelNames, metricNameAndLabelValuesAndValueList)))
+                            .orElseGet(() -> Stream
+                                    .of(buildGaugeWithoutLabels(metricName, help, metricNameAndLabelValuesAndValueList.stream().map(x -> x.value).iterator().next())));
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:54]: It is not :) The code should also check if the list has only one element, refactored this into `firstElementFromSingletonListOrThrow`.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            return result;
+        }
+    }
+
+    private static MetricNameAndLabelValuesAndValue metricNameAndLabelValuesAndValueForReportEntry(ReportEntry reportEntry,
+                                                                                                   Set<String> metricNames) {
+        String rawMetricName = reportEntry.getKey().toString().replace(REPORT_ENTRY_KEY_SEP, METRIC_NAME_SEP);
+        return metricNames.stream()
+                .filter(rawMetricName::startsWith)
+                .findAny()
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:95]: It was indeed not documented, but the whole algorithm of metric extraction is refactored to take into account IIS key-to-metric mapping cases and labeled metric recognition is now done using regexes.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                                            reportLocationsFinder.find(parameters)
+                                                    .ifPresent(paths -> {
+                                                        logger.info("Using report locations of size: {}", paths.size());
+                                                        labelNamesByMetricNameProducer.create(parameters)
+                                                                .ifPresent(labelNamesByMetricName -> {
+                                                                    logger.info("Using label map of size: {}", labelNamesByMetricName.size());
+                                                                    List<Optional<List<Gauge>>> optionals = paths.stream()
+                                                                            .map(path -> reportEntryReader.read(fs, new Path(path))
+                                                                                    .flatMap(reportEntries -> reportEntryConverter.convert(reportEntries, path, labelNamesByMetricName)))
+                                                                            .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java:77]: Note: This part of code is refactored but the comment is still valid. It would seem like a good solution but I think the better solution is to stay with `Optional`. In our new approach we specify as input for the reader a dir containing reports as subdirs. The dir is the same for all workflows, including the one that generates summary json report. But reading json report using avro reader fails. We can cleanly support this situation with non-empty optional for successful reading and empty optional for unsuccessful reading.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java (hunk 1 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java:32]: Changes
1. use `push`
1. job is always `iis`
1. added grouping key map with keys and values supplied from workflow; currently only one grouping key `user`

**Comment by przemyslawjacewicz** [comment]
> A hard case for better labelled metrics. It would be very nice to have:
> 
>     * `import.content.url.bytype.pdf` => `import_content_url{type="pdf"}`
> 
>     * `import.content.pdf.correct` => `import_content{type="pdf",status="correct"}`
> 
>     * `import.content.pdf.invalid.sizeExceeded` => `import_content{type="pdf",status="sizeExceeded"}`
> 
> 
> I can imagine configuration here in the form of a mapping (ordered by prefix length) _prefix => [(label, suffix_extractor), ...]_, eg.:
> 
>     1. _import.content.url => [(type, (xs) -> xs[1])]_
> 
>     2. _import.content => [(type, (xs) -> xs[0]), (status, (xs) -> xs[xs.length-1])]_

To support the harder case we need to use bigger guns because simple string matching won't work. So the key-to-metric mapping is done using regexes with configuration in properties file. The sample configuration looks like this:
```bash
a\\.b\\.c\\.d\\.\\S+={"metricName":"a_b_c_d","labelConfs":[{"labelName":"label_name","pattern":"$4"}]}
```
The part on the left hand side is a regex that should match to report key we want to convert into a labeled metric. Unfortunately we must escape dots and regex tokens. The part on the right hand side is a json representing a new class `LabeledMetricConf` containing the configuration of the metric and labels. Label values are extracted using positional placeholder `$pos`, with position starting from 0 and representing the index of an element of an array constructed by splitting the key using dots. This configuration supports concatenations like `$3_$4`. This type of labeled metric configuration should support all key to metric mapping currenlty in iis.

File: iis-wf/iis-wf-report/src/test/resources/eu/dnetlib/iis/wf/report/pushgateway/process/oozie_app/workflow.xml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    </action>
+
+    <action name="push-reports">
+        <java>
+            <main-class>eu.dnetlib.iis.common.java.ProcessWrapper</main-class>
+            <arg>eu.dnetlib.iis.wf.report.pushgateway.process.PushMetricsProcess</arg>
+            <arg>-PjobName=test-pushgateway-push</arg>
+            <arg>-PmetricPusherCreatorClassName=eu.dnetlib.iis.wf.report.pushgateway.pusher.FilePusherCreator</arg>
+            <arg>-PmetricPusherAddress=${workingDir}/output</arg>
+            <arg>-PlabeledMetric.processing_documentClassification_docClassReference=type</arg>
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/test/resources/eu/dnetlib/iis/wf/report/pushgateway/process/oozie_app/workflow.xml:44]: As discussed this type of labeled metric configuration is removed and is substituted by configuration in properties file. The configuration there uses regexes to filter report keys that should be converted to labeled metric.

File: iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/export/oozie_app/workflow.xml [hunk 6]
```diff
@@ -235,6 +235,21 @@
...
         </property>
+        <property>
+            <name>metric_pusher_creator_class_name</name>
+            <value>eu.dnetlib.iis.wf.report.pushgateway.pusher.PushGatewayPusherCreator</value>
+            <description>
+                canonical name of a class implementing eu.dnetlib.iis.wf.report.pushgateway.pusher.MetricPusherCreator
+                used to create eu.dnetlib.iis.wf.report.pushgateway.pusher.MetricPusher for pushing prometheus metrics
+                to pushgateway
+            </description>
+        </property>
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/export/oozie_app/workflow.xml:246]: Done, as discussed before, the new push reports workflow for non-test runs needs only one parameter `reportsDirPath` with the location of reports to be pushed. This is the full list of parameters for the workflow:
1. `metric_pusher_creator_class_name` -> with default value
1. `metric_pusher_address` -> with default value
1. `reports_dir_path`
1. `labeled_metrics_properties_file` -> with default value

File: iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/import/oozie_app/workflow.xml [hunk 7]
```diff
@@ -775,10 +790,48 @@
...
+			<arg>-PmetricPusherCreatorClassName=${metric_pusher_creator_class_name}</arg>
+			<arg>-PmetricPusherAddress=${metric_pusher_address}</arg>
+			<arg>-PlabeledMetric.import_content_html_invalid=cause</arg>
+			<arg>-PlabeledMetric.import_content_pdf_invalid=cause</arg>
+			<arg>-PlabeledMetric.import_content_pmc_invalid=cause</arg>
+			<arg>-PlabeledMetric.import_content_url_bytype=type</arg>
+			<arg>-PlabeledMetric.import_content_wos_invalid=cause</arg>
+			<arg>-PlabeledMetric.import_infoSpace=type</arg>
+			<arg>-PlabeledMetric.import_metadataExtraction_processed_fault=cause</arg>
+			<arg>-PreportLocation.import_concept=${output_report_root_path}/import_concept</arg>
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/import/oozie_app/workflow.xml:811]: We also agreed to use `post` method, which in terms of prometheus api is `push'

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Converts report entries to prometheus gauges.
+ * <p>
+ * For report entries containing durations gauge's name is created from report key by replacing dots with
+ * underscores and adding '_seconds' suffix. For report entries containing counters gauge's name is created from report
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:15]: Yes, I thought is was more or less implied ;-) Note: due to converter changes this doc was split and moved to classes responsible for converting counters and durations separately.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+                                                                                                        metricPusher.pushSafe(registry, jobName);
+                                                                                                    });
+                                                                                        });
+                                                                            });
+                                                                });
+                                                    });
+                                        });
+                            });
+                });
+    }
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java:97]: Yes, the code goes to the right side of the screen pretty fast :/ The usage of `Optional' seemed to fit for the case. Formatting could be probably changed using some custom setting, here I'm using default IDEA's formatting. We could even introduce project-wise code formatting rules which would remove this kind of problems.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+    public static class MetricPusherCreatorProducer {
+        Optional<MetricPusherCreator> create(Map<String, String> parameters) {
+            Supplier<MetricPusherCreator> metricPusherCreatorSupplier = () -> {
+                try {
+                    return (MetricPusherCreator) Class.forName(parameters.get(METRIC_PUSHER_CREATOR_CLASS_NAME))
+                            .getConstructor().newInstance();
+                } catch (Exception e) {
+                    logger.error("MetricPusherCreator creation failed.");
+                    e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java:122]: I think it was autogenerated at some point and I blindly copy-pasted.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        private static String metricNameOrThrow(String parameterName) {
+            return parameterName.substring(LABELED_METRIC.length() + 1);
+        }
+
+        private static String[] labelNamesOrThrow(String parameterValue) {
+            String[] labelNames = parameterValue.split(METRIC_LABEL_NAMES_SEP);
+            if (labelNames.length == 0 || StringUtils.isAnyBlank(labelNames)) {
+                throw new RuntimeException("Label names creation failed.");
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java:233]: Due to label extraction refactorization this exception throw statement is removed. But the corresponding method, which parses label metric conf from json, has additional dupm of failing configuration, check out `PushMetricsProcess.LabeledMetricConfByPatternProducer::labeledMetricConfOrThrow`

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+                    } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/pusher/PushGatewayPusherCreator.java:33]: Yes, you are. However, the only scenario I could test is a failure of pushgateway service.

**Comment by przemyslawjacewicz** [comment]
> I am done.
> 
> Most of the issues were already discussed during today's call (I tried to add appropriate remarks in comments) including the most important:
> 
>     * relying on the same root report dir in all 3 "pusher" references
> 
>     * moving all the label definitions into single `.properties` file (referenced in all "pusher" instances), we should probably reference this file and describe it on the [IIS reports](https://issue.openaire.research-infrastructures.eu/projects/openaire/wiki/IIS_reports) wiki page
> 
>     * including `${user.name}` context when exporting metrics (so we have different environments beta and prod distinguished)
> 
> 
> One thing we didn't discuss and I find pretty important is the knowledge about any failures happening during the export procedure. It seems we are immune to any kind of failures but we should make ourselves aware of those failures whenever they happen. I guess we could generate dedicated report entry (entries?) counting the unsuccessful tries.

I've implemented all the changes discussed during our call. The only thing that seems to be left out is how to report the status of pushing reports. I think we settled down to manual check in grafana for any missing metrics. If this will resolve to be a problematic solution we could add some additional reporting.

**Comment by przemyslawjacewicz** [comment]
Note: the mechanism of label extraction was refactored significantly. This corresponds major changes in `eu.dnetlib.iis.wf.report.pushgateway.converter` package.

File: iis-wf/iis-wf-report/src/test/java/eu/dnetlib/iis/wf/report/pushgateway/process/FilePusherCreator.java [hunk 13]
```diff
@@ -22,15 +24,15 @@
...
             List<ReportEntry> reportEntries = Collections.list(collectorRegistry.metricFamilySamples()).stream()
                     .flatMap(metricFamilySamples -> metricFamilySamples.samples.stream().map(FilePusherCreator::sampleToReportEntry))
                     .collect(Collectors.toList());
-
             try {
                 FileSystem fs = FileSystem.get(new Configuration());
-                DataStore.create(reportEntries, new FileSystemPath(fs, new Path(address, jobName)), ReportEntry.SCHEMA$);
+                DataStore.create(reportEntries, new FileSystemPath(fs, new Path(address, job)), ReportEntry.SCHEMA$);
             } catch (IOException e) {
+                //TODO log exception
```
Comment (marekhorst) [inline iis-wf/iis-wf-report/src/test/java/eu/dnetlib/iis/wf/report/pushgateway/process/FilePusherCreator.java:35]: Let's get rid of TODOs and let's log the exception by logger.

File: iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java [hunk 14]
```diff
@@ -1,37 +1,53 @@
...
-    
+
+    /**
+     * Lists hadoop files located below pathname or alternatively lists subdirs under pathname.
+     *
+     * @param hadoopConf Configuration of hadoop env
+     * @param pathname   Path to be listed for hadoop files
+     * @return List with string locations of hadoop files
+     */
+    public static List<String> listFiles(Configuration hadoopConf, String pathname) throws IOException {
```
Comment (marekhorst) [inline iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:45]: Does the logic within this method reflect the description? It seems it lists directories only (filtering by `FileStatus::isDirectory`).

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+    public static Gauge buildGaugeWithLabels(String metricName,
+                                             String help,
+                                             List<String> labelNames,
+                                             List<List<String>> labelValues,
+                                             List<Double> values) {
+        Gauge gauge = Gauge.build()
+                .name(metricName)
+                .help(help)
+                .labelNames(labelNames.toArray(new String[0]))
```
Comment (marekhorst) [inline iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java:30]: I guess it will be optimal to initialize the `String` array with the known size (`labelNames#size()`).

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java [hunk 16]
```diff
@@ -4,151 +4,116 @@
...
-                                                 String help,
-                                                 double value) {
-        Gauge gauge = Gauge.build()
-                .name(metricName)
-                .help(help)
-                .create();
-        gauge.set(value);
-        return gauge;
+    private static <E, X> X firstElementFromSingletonListOrThrow(List<E> list, Function<E, X> mapper) {
+        List<X> collect = list.stream().map(mapper).distinct().collect(Collectors.toList());
```
Comment (marekhorst) [inline iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:112]: Hmm, do we have to make stream out of the list to take the first element?

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+                .create();
+        ListUtils.zip(labelValues, values).forEach(labelValuesAndValue ->
+                gauge.labels(labelValuesAndValue.getLeft().toArray(new String[0])).set(labelValuesAndValue.getRight())
```
Comment (marekhorst) [inline iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java:33]: Same here, we could initialize `String` array with an expected size.

**Comment by marekhorst** [comment]
Forgot to add review comment: I've just finished code review.

File: iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/import/oozie_app/workflow.xml [hunk 18]
```diff
@@ -783,7 +772,7 @@
             <arg>eu.dnetlib.iis.common.report.OozieTimeReportGenerator</arg>
             <arg>-PjobId=${wf:id()}</arg>
             <arg>-PoozieServiceLoc=${oozieServiceLoc}</arg>
-            <arg>-Preport.import.concept.duration=import_concept</arg>
+            <arg>-Preport.import.concepts.duration=import_concept</arg>
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-primary/src/main/resources/eu/dnetlib/iis/wf/primary/import/oozie_app/workflow.xml:775]: The changes leaked by accident :( I've restored the previous version for `import.concept` and for `import.content.html`.

File: iis-wf/iis-wf-report/src/test/java/eu/dnetlib/iis/wf/report/pushgateway/process/FilePusherCreator.java (hunk 13 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/test/java/eu/dnetlib/iis/wf/report/pushgateway/process/FilePusherCreator.java:35]: Missed that one, heh

File: iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java (hunk 14 above)
Comment (przemyslawjacewicz) [reply iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:45]: By 'hadoop files' I meant dirs with `part` files inside. This is actually a copy-paste from `dnet-hadoop`. I'll change the name to `listDirs` and description to say that the method lists subdirs of a given dir.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java (hunk 15 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java:30]: I did that, but IDE flagged this and suggested to convert it to the version with `new String[0]`. Here's the description IDE gives:

Call to 'toArray()' with pre-sized array argument 'new String[labelNames.size()]' 
 Inspection info: There are two styles to convert a collection to an array: either using a pre-sized array (like c.toArray(new String[c.size()])) or using an empty array (like c.toArray(new String[0]).
In older Java versions using pre-sized array was recommended, as the reflection call which is necessary to create an array of proper size was quite slow. However since late updates of OpenJDK 6 this call was intrinsified, making the performance of the empty array version the same and sometimes even better, compared to the pre-sized version. Also passing pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray call which may result in extra nulls at the end of the array, if the collection was concurrently shrunk during the operation.
This inspection allows to follow the uniform style: either using an empty array (which is recommended in modern Java) or using a pre-sized array (which might be faster in older Java versions or non-HotSpot based JVMs).

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java (hunk 16 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:112]: Generally no, but in this case, apart from applying also a mapper, we want to check that this the list contains only one unique element.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java (hunk 17 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/GaugesBuilder.java:33]: https://github.com/openaire/iis/pull/1072#discussion_r426593666

File: iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java [hunk 19]
```diff
@@ -1,37 +1,53 @@
...
+
     /**
      * Removes file or directory (recursively) located under the specified pathname.
      */
     public static void remove(Configuration hadoopConf, String pathname) throws IOException {
-        
         Path path = new Path(pathname);
-        
         FileSystem fileSystem = FileSystem.get(hadoopConf);
-
```
Comment (mpol) [inline iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:30]: Including such unrelated formatting changes makes the review a bit harder than it would be otherwise.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java (hunk 16 above)
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/ReportEntryToMetricConverter.java:112]: To make this more visible I've changed the name of this method to `uniqueElementOrThrow`.

File: iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java (hunk 19 above)
Comment (przemyslawjacewicz) [reply iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:30]: Yes, it also adds to pull request's changed lines value. It's an effect of applying IDE's code formatting for the entire file, not only for the code that's added.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+                LabelValuesExtractor::extractLabelValuesByPosition);
+    }
+
+    public static ExtractedMetric extractMetricFromReportEntry(ReportEntry reportEntry,
+                                                               Map<String, LabeledMetricConf> labeledMetricConfByPattern,
+                                                               MetricNameCustomizer.Customizer metricNameCustomizer,
+                                                               LabelValuesExtractor.Extractor labelValuesExtractor) {
+        return labeledMetricConfByPattern.entrySet().stream()
+                .filter(entry -> keyMatchesPattern(reportEntry.getKey().toString(), entry.getKey()))
+                .findFirst()
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java:34]: For labeled metric report entry's key should match only one regex from the map keys. Then the map will contain only one element and I've used `findFirst` to get that one entry. It could also be `findAny`. However, it is confusing so I've rewritten this code: first filer map entries matching report's key, then if there is an entry check if there's only one entry and extract metric using label metric conf; if there's no match return metric without labels.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+        return Pattern.compile(addStartEndAnchors(keyPattern)).matcher(key).find();
+    }
+
+    private static String addStartEndAnchors(String pattern) {
+        return String.format("^%s$", pattern);
+    }
+
+    private static String defaultMetricName(String key) {
+        return key.replace(ReportEntryToMetricConverter.REPORT_ENTRY_KEY_SEP, ReportEntryToMetricConverter.METRIC_NAME_SEP);
+    }
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java:58]: Original idea was that we may have to apply more than one metric naming conventions forced by prometheus api. However, we only have one, metric name cannot have dashes, as for example the key for covid minining containing `covid-19` string. So separate name customization is unnecessary and I'll move its functionality to the code replacing periods with underscores.

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+        return extractMetricFromReportEntry(reportEntry,
+                labeledMetricConfByPattern,
+                MetricNameCustomizer::dashRemover,
+                LabelValuesExtractor::extractLabelValuesByPosition);
+    }
+
+    public static ExtractedMetric extractMetricFromReportEntry(ReportEntry reportEntry,
+                                                               Map<String, LabeledMetricConf> labeledMetricConfByPattern,
+                                                               MetricNameCustomizer.Customizer metricNameCustomizer,
+                                                               LabelValuesExtractor.Extractor labelValuesExtractor) {
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtractor.java:31]: I didn't see that naming pattern before but you're right, the repetitions are unnecessary. `Function` seems too general, though, so maybe a change in the name of the container class but leave the name of the functional interface: `CounterReportEntryMetricExtraction.Extractor` ?

File: iis-wf/iis-wf-report/src/main/resources/eu/dnetlib/iis/wf/report/pushgateway/process/oozie_app/labeled_metrics.properties [hunk 23]
```diff
@@ -0,0 +1,16 @@
+# import
+import\\.content\\.\\S+\\.correct={"metricName":"import_content","labelConfs":[{"labelName":"type","pattern":"$2"},{"labelName":"status","pattern":"$3"}]}
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/resources/eu/dnetlib/iis/wf/report/pushgateway/process/oozie_app/labeled_metrics.properties:2]: I've added the description on redmine [Prometheus metrics](https://issue.openaire.research-infrastructures.eu/projects/openaire/wiki/IIS_reports#Prometheus-metrics). The whole page on IIS reports will need to be updated also after the change in report keys.

**Comment by marekhorst** [comment]
It looks good to me, thank's Przemek for the fixes!

File: iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtraction.java [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+        return labeledMetricConf.toExtractedMetric(labelValues, value);
+    }
+
+    private static Boolean keyMatchesPattern(String key, String keyPattern) {
+        return Pattern.compile(addStartEndAnchors(keyPattern)).matcher(key).matches();
+    }
+
+    private static String addStartEndAnchors(String pattern) {
+        return String.format("^%s$", pattern);
+    }
```
[the remark this answers is not in the document]
Comment (przemyslawjacewicz) [reply iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/converter/CounterReportEntryMetricExtraction.java:74]: Right, didn't check the difference between `Matcher#matches()` and `Matcher#find()`.

**Comment by przemyslawjacewicz** [comment]
One more fix: removed unnecessary anchors addition.
